src/status_im/chat/events/receive_message.cljs

```diff
@@ -16,30 +17,49 @@
  (fn [cofx [message]]
    (message-model/receive message cofx)))
 
+(re-frame.core/reg-fx
+ :chat-received-message/add-fx
+ (fn [messages]
+   (re-frame/dispatch [:chat-received-message/add messages])))
+
+(defn add-one-message [[{:keys [content] :as message}] {:keys [db] :as cofx}]
+  (when (message-model/add-to-chat? cofx message)
+    (if (:command content)
+      ;; we are dealing with received command message, we can't add it right away,
+      ;; we first need to fetch short-preview + preview and add it only after we already have those.
+      ;; note that `request-command-message-data` implicitly wait till jail is ready and
+      ;; calls are made only after that
+      (commands-events/request-command-message-data
+       db message
+       {:data-type             :short-preview
+        :proceed-event-creator (fn [short-preview]
+                                 [:request-command-message-data
+                                  message
+                                  {:data-type             :preview
+                                   :proceed-event-creator (fn [preview]
+                                                            [::received-message
+                                                             (update message :content merge
+                                                                     {:short-preview short-preview
+                                                                      :preview       preview})])}])})
+      ;; regular non command message, we can add it right away
+      (message-model/receive message cofx))))
+
+(defn add [[messages] {:keys [db] :as cofx}]
+  (reduce (fn [fx message]
+            (let [temp-cofx (handlers-macro/update-db cofx fx)]
+              (handlers-macro/safe-merge
+               fx
+               (add-one-message [message] temp-cofx))))
```

Comment:
We can't just reduce effects produced by `add-one-message`:
What if there are multiple command type messages in the batch and all needs their preview to be rendered?
What if the messages are for currently opened chat and we need to produce protocol seen messages for all of them ?
What about updating adding request objects to realm (currently, there is no plural effect for that, till my `realm-transactions-2` PR is merged).
Thats why I think we need to tackle this differently.

Reply:
> What if there are multiple command type messages in the batch and all needs their preview to be rendered?
we can always "accumulate" these preview requests here, not a big deal

> What if the messages are for currently opened chat and we need to produce protocol seen messages for all of them ?
same answer, if we need to call some fx it can be batched as well

> What about updating adding request objects to realm (currently, there is no plural effect for that, till my realm-transactions-2 PR is merged).
so we will merge it, anyway

Reply:
> What if the messages are for currently opened chat and we need to produce protocol seen messages for all of them ?

tbh i even doubt that this makes sense in case if user opened public group after starting app and then received like 100 messages from mailserver or so. Why these messages should be marked as seen? That doesn't make sense, the fact that chat was opened doesn't mean that messages were seen. I believe this should be handled completely differently considering UX pov.

Reply:
@rasom We don't even have send confirmations for public chats anymore (@cammellos removed them recently). But the same situation can happen in 1-1 chat (going online after longer time and receiving bunch of messages from offline inboxing) and there, it makes perfect sense to send just one seen protocol message with N confirmations, instead of N protocol messages (whisper envelopes are expensive).
Regarding if it makes sense to mark messages as seen upon opening the chat, I don't know how to otherwise detect it, as the condition that chat is opened and new message arrived is synonymous with rendering the message in messages list. The only alternative I see is having some explicit confirmation or much more aggressive lazy rendering when only what's in the content of the screen is really rendered (and hooking seen confirmation into that).

Reply:
>if you actually seen them and not just received. But we can't know this when handling message.

Sure you can - any good design should be data first. Currently we know that because of the `:current-chat-id` in db and `:chat-id` of the message, in the future, with lazy rendering (the change you are talking about), what's rendered will be determined by some range like `:rendered-messages [from to]` and you will now if the new message is destined for the currently opened chat and will be included in rendered range.

Reply:
and also we need to take into account how long is message, how big is screen and is keyboard shown or not. I still doubt that this makes sense on handling, though it's doable, I agree :D